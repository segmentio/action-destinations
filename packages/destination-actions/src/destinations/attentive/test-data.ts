import { SegmentEvent } from '@segment/actions-core' // Adjust the import as needed for your project

// Define the function with a return type for clarity
export function getCustomEventsTestValidPayload(timestamp: string): Partial<SegmentEvent> {
  return {
    timestamp,
    event: 'Event Type 1',
    messageId: '123e4567-e89b-12d3-a456-426614174000',
    type: 'track',
    userId: '123e4567-e89b-12d3-a456-426614174000',
    context: {
      traits: {
        phone: '+3538675765689',
        email: 'test@test.com'
      }
    },
    properties: {
      tracking_url: 'https://tracking-url.com',
      product_name: 'Product X'
    }
  }
}

export function getCustomEventsTestMapping() {
  return {
    type: { '@path': '$.event' },
    userIdentifiers: {
      phone: { '@path': '$.context.traits.phone' },
      email: { '@path': '$.context.traits.email' },
      clientUserId: { '@path': '$.userId' }
    },
    properties: { '@path': '$.properties' },
    externalEventId: { '@path': '$.messageId' },
    occurredAt: { '@path': '$.timestamp' }
  }
}

export function getCustomEventsTestExpectedPayload(validPayload: Record<string, any>) {
  return {
    type: validPayload.type,
    properties: validPayload.properties,
    externalEventId: validPayload.messageId,
    occurredAt: validPayload.timestamp,
    user: {
      phone: validPayload.context?.traits?.phone,
      email: validPayload.context?.traits?.email,
      externalIdentifiers: {
        clientUserId: validPayload.userId
      }
    }
  }
}

export function getECommEventTestValidPayload(timestamp: string): Partial<SegmentEvent> {
  return {
    timestamp,
    event: 'Product Viewed',
    messageId: '123e4567-e89b-12d3-a456-426614174001',
    type: 'track',
    userId: '123e4567-e89b-12d3-a456-426614174001',
    context: {
      traits: {
        phone: '+3538675765689',
        email: 'test@test.com'
      }
    },
    properties: {
      items: [
        {
          productId: 'prod_123',
          productVariantId: 'var_456',
          productImage: 'https://image-url.com/product.jpg',
          productUrl: 'https://product-url.com',
          name: 'Product X',
          price: {
            value: 29.99,
            currency: 'USD'
          },
          quantity: 1
        }
      ]
    }
  }
}

export function getECommEventTestMapping() {
  return {
    items: [
      {
        productId: { '@path': '$.properties.items[0].productId' },
        productVariantId: { '@path': '$.properties.items[0].productVariantId' },
        productImage: { '@path': '$.properties.items[0].productImage' },
        productUrl: { '@path': '$.properties.items[0].productUrl' },
        name: { '@path': '$.properties.items[0].name' },
        price: {
          value: { '@path': '$.properties.items[0].price.value' },
          currency: { '@path': '$.properties.items[0].price.currency' }
        },
        quantity: { '@path': '$.properties.items[0].quantity' }
      }
    ],
    userIdentifiers: {
      phone: { '@path': '$.context.traits.phone' },
      email: { '@path': '$.context.traits.email' },
      clientUserId: { '@path': '$.userId' }
    },
    occurredAt: { '@path': '$.timestamp' }
  }
}

export function getECommEventTestExpectedPayload(validPayload: Record<string, any>) {
  return {
    items: validPayload.properties?.items,
    occurredAt: validPayload.timestamp,
    user: {
      phone: validPayload.context?.traits?.phone,
      email: validPayload.context?.traits?.email,
      externalIdentifiers: {
        clientUserId: validPayload.userId
      }
    }
  }
}
