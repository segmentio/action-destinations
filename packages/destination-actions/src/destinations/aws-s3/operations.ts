import { ExecuteInput } from '@segment/actions-core'
import type { Payload } from './syncAudienceToCSV/generated-types'
import type { AudienceSettings } from './generated-types'

// Type definitions
export type RawData = {
  context?: {
    personas?: {
      computation_key?: string
      computation_class?: string
      computation_id?: string
    }
  }
}

export type ExecuteInputRaw<Settings, Payload, RawData, AudienceSettings = unknown> = ExecuteInput<
  Settings,
  Payload,
  AudienceSettings
> & { rawData?: RawData }

function generateFile(payloads: Payload[], audienceSettings: AudienceSettings): string {
  const headers: string[] = []
  const columnsField = payloads[0].columns
  const additionalColumns = payloads[0].additional_identifiers_and_traits_columns ?? []

  Object.entries(columnsField).forEach(([_, value]) => {
    if (![undefined, null, ''].includes(value)) {
      headers.push(value)
    }
  })

  additionalColumns.forEach((additionalColumn) => {
    if (![undefined, null, ''].includes(additionalColumn.value)) {
      headers.push(additionalColumn.value)
    }
  })

  const headerString = `${headers.join(audienceSettings.delimiter === 'tab' ? '\t' : audienceSettings.delimiter)}\n`
  const rows: string[] = [headerString]

  payloads.forEach((payload, index, arr) => {
    const action = payload.propertiesOrTraits[payload.audienceName]

    const row: string[] = []
    if (![undefined, null, ''].includes(columnsField.audience_name)) {
      row.push(encodeString(String(payload.audienceName ?? '')))
    }
    if (![undefined, null, ''].includes(columnsField.audience_id)) {
      row.push(encodeString(String(payload.audienceId ?? '')))
    }
    if (![undefined, null, ''].includes(columnsField.audience_action)) {
      row.push(encodeString(String(action ?? '')))
    }
    if (![undefined, null, ''].includes(columnsField.email)) {
      row.push(encodeString(String(payload.email ?? '')))
    }
    if (![undefined, null, ''].includes(columnsField.user_id)) {
      row.push(encodeString(String(payload.userId ?? '')))
    }
    if (![undefined, null, ''].includes(columnsField.anonymous_id)) {
      row.push(encodeString(String(payload.anonymousId ?? '')))
    }
    if (![undefined, null, ''].includes(columnsField.timestamp)) {
      row.push(encodeString(String(payload.timestamp ?? '')))
    }
    if (![undefined, null, ''].includes(columnsField.message_id)) {
      row.push(encodeString(String(payload.messageId ?? '')))
    }
    if (![undefined, null, ''].includes(columnsField.space_id)) {
      row.push(encodeString(String(payload.spaceId ?? '')))
    }
    if (![undefined, null, ''].includes(columnsField.integrations_object)) {
      row.push(encodeString(String(JSON.stringify(payload.integrationsObject) ?? '')))
    }
    if (![undefined, null, ''].includes(columnsField.properties_or_traits)) {
      row.push(encodeString(String(JSON.stringify(payload.propertiesOrTraits) ?? '')))
    }

    additionalColumns.forEach((additionalColumn) => {
      row.push(encodeString(String(JSON.stringify(payload.propertiesOrTraits[additionalColumn.key]) ?? '')))
    })

    const isLastRow = arr.length === index + 1
    const rowString = `${row.join(audienceSettings.delimiter === 'tab' ? '\t' : audienceSettings.delimiter)}${
      isLastRow ? '' : '\n'
    }`

    rows.push(rowString)
  })

  return rows.join('')
}

function encodeString(str: string) {
  return `"${String(str).replace(/"/g, '""')}"`
}

function validate(payloads: Payload[], audienceSettings: AudienceSettings) {
  const delimiter = audienceSettings.delimiter
  const columns = payloads[0].columns
  const additionalIdentifierColumns = payloads[0].additional_identifiers_and_traits_columns

  // ensure column names do not contain delimiter
  Object.values(columns).forEach((columnName) => {
    if (columnName.includes(delimiter)) {
      throw new Error(`Column name ${columnName} cannot contain delimiter: ${delimiter}`)
    }
  })

  // ensure additional identifier column names do not contain delimiter
  additionalIdentifierColumns?.forEach((column) => {
    if (column.value.includes(delimiter)) {
      throw new Error(`Column name ${column.value} cannot contain delimiter: ${delimiter}`)
    }
  })
}

export { generateFile, validate }
