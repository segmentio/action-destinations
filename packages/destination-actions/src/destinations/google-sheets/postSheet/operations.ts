import type { Payload } from './generated-types'
import { RequestClient } from '@segment/actions-core'
import { GoogleSheets, GetResponse } from '../googleapis/index'

import A1 from '@segment/a1-notation'

type Fields = {
  [k: string]: string
}

/**
 * Invariant settings that are common to all events in the payload.
 */
export type MappingSettings = {
  spreadsheetId: string
  spreadsheetName: string
  dataFormat: string
  columns: string[]
}

/**
 * Utility function that converts the event properties into an array of strings that Google Sheets API can understand.
 * Note that the identifier is forced as the first column. 
 * @param identifier value used to imbue fields with a uniqueness constraint
 * @param fields list of properties contained in the event
 * @param columns list of properties that will be committed to the spreadsheet
 * @returns a string object that has used the `fields` data to populate the `columns` ordering
 * 
 * @example
 * fields: 
    {
      "CLOSE_DATE": "2022-07-08T00:00:00Z",
      "CLOSE_DATE_EOQ": "2022-07-08",
      "ENTRY_POINT": "Website Demo Request",
      "E_ARR_POST_LAUNCH_C": "100000.0",
      "FINANCE_ENTRY_POINT": "Inbound High Intent"
    } 
    columns: ["ENTRY_POINT", "MISSING_COLUMN", "CLOSE_DATE"]

    return => ["Website Demo Request", "", "2022-07-08T00:00:00Z"]

 */
const generateColumnValuesFromFields = (identifier: string, fields: Fields, columns: string[]) => {
  const retVal = columns.map((col) => fields[col] ?? '')
  retVal.unshift(identifier) // Write identifier as first column
  return retVal
}

/**
 * Processes the response of the Google Sheets GET call and parses the events into separate operation buckets.
 * @param response result of the Google Sheets API get call
 * @param events data to be written to the spreadsheet
 * @returns
 */
function processGetSpreadsheetResponse(response: GetResponse, events: Payload[]) {
  // TODO (STRATCONN-1375): Fail request if above row limit

  const updateBatch: { identifier: string; event: Fields; targetIndex: number }[] = []
  const appendBatch: { identifier: string; event: Fields }[] = []

  // Use a hashmap to efficiently find if the event already exists in the spreadsheet (update) or not (append).
  const eventMap = new Map(events.map((e) => [e.record_identifier, e]))

  if (response.values && response.values.length > 0) {
    for (let i = 1; i < response.values.length; i++) {
      const targetIdentifier = response.values[i][0]
      if (eventMap.has(targetIdentifier)) {
        // The event being processed already exists in the spreadsheet.
        const targetEvent = eventMap.get(targetIdentifier) as Payload
        if (targetEvent.operation_type != 'deleted') {
          updateBatch.push({
            identifier: targetIdentifier,
            event: targetEvent.fields as Fields,
            targetIndex: i + 1
          })
        }
        eventMap.delete(targetIdentifier)
      }
    }
  }

  // At this point, eventMap contains all the rows we couldn't find in the spreadsheet.
  eventMap.forEach((value, key) => {
    // If delete, just drop event
    if (value.operation_type != 'deleted') {
      appendBatch.push({
        identifier: key,
        event: value.fields as Fields
      })
    }
  })

  return { appendBatch, updateBatch }
}

/**
 * Commits all passed events to the correct row in the spreadsheet, as well as the columns header row.
 * @param mappingSettings configuration object detailing parameters for the call
 * @param updateBatch array of events to commit to the spreadsheet
 * @param gs interface object capable of interacting with Google Sheets API
 */
async function processUpdateBatch(
  mappingSettings: MappingSettings,
  updateBatch: { identifier: string; event: { [k: string]: string }; targetIndex: number }[],
  gs: GoogleSheets
) {
  // Utility function used to calculate which range an event should be written to
  const getRange = (targetIndex: number, columnCount: number) => {
    const targetRange = new A1(1, targetIndex)
    targetRange.addX(columnCount)
    return targetRange.toString()
  }

  const batchPayload = updateBatch.map(({ identifier, event, targetIndex }) => {
    // Flatten event fields to be just the values
    const values = generateColumnValuesFromFields(identifier, event, mappingSettings.columns)
    return {
      range: `${mappingSettings.spreadsheetName}!${getRange(targetIndex, values.length)}`,
      values: [values]
    }
  })

  // Always add to the payload a write to the first row (containing column names) in case that columns have been updated
  const headerRowRange = new A1(1, 1, 1, mappingSettings.columns.length + 1)
  batchPayload.push({
    range: `${mappingSettings.spreadsheetName}!${headerRowRange.toString()}`,
    values: [['id', ...mappingSettings.columns]]
  })

  return gs
    .batchUpdate(mappingSettings, batchPayload)
    .then(() => {
      console.log('update')
    })
    .catch((error) => {
      console.log(error)
    })
}

// TODO: Re-enable delete once locking is supported.
/**
 * Clears all passed events from the spreadsheet.
 * @param mappingSettings configuration object detailing parameters for the call
 * @param updateBatch array of events to clear from the spreadsheet
 * @param gs interface object capable of interacting with Google Sheets API
 */
// async function processDeleteBatch(
//   mappingSettings: MappingSettings,
//   deleteBatch: { identifier: string; targetIndex: number }[],
//   gs: GoogleSheets
// ) {
//   if (deleteBatch.length <= 0) {
//     return
//   }

//   // TODO: fix a1-notation package to support 1:1 notation
//   const deletePayload = deleteBatch.map(({ targetIndex }) => {
//     return {
//       range: `${mappingSettings.spreadsheetName}!${targetIndex}:${targetIndex}`
//     }
//   })

//   return gs
//     .batchClear(mappingSettings, deletePayload)
//     .then(() => {
//       console.log('delete')
//     })
//     .catch((error) => {
//       console.log(error)
//     })
// }

/**
 * Commits all passed events to the bottom of the spreadsheet.
 * @param mappingSettings configuration object detailing parameters for the call
 * @param appendBatch array of events to commit to the spreadsheet
 * @param gs interface object capable of interacting with Google Sheets API
 * @returns
 */
async function processAppendBatch(
  mappingSettings: MappingSettings,
  appendBatch: { identifier: string; event: { [k: string]: string } }[],
  gs: GoogleSheets
) {
  if (appendBatch.length <= 0) {
    return
  }

  // Flatten event fields to be just the values
  const values = appendBatch.map(({ identifier, event }) =>
    generateColumnValuesFromFields(identifier, event, mappingSettings.columns)
  )

  // Start from A2 to skip header row (in case it has not been written yet)
  return gs
    .append(mappingSettings, 'A2', values)
    .then(() => {
      console.log('append')
    })
    .catch((error) => {
      console.log(error)
    })
}

/**
 * Takes an array of events and dynamically decides whether to append, update or delete rows from the spreadsheet.
 * @param request request object used to perform HTTP calls
 * @param events array of events to commit to the spreadsheet
 */
function processData(request: RequestClient, events: Payload[]) {
  // These are assumed to be constant across all events
  const mappingSettings = {
    spreadsheetId: events[0].spreadsheet_id,
    spreadsheetName: events[0].spreadsheet_name,
    dataFormat: events[0].data_format,
    columns: Object.getOwnPropertyNames(events[0].fields)
  }

  const gs: GoogleSheets = new GoogleSheets(request)

  // Get all of the row identifiers (assumed to be in the first column A)
  gs.get(mappingSettings, 'A:A')
    .then((response) => {
      // Use the retrieved row identifiers along with the incoming events to decide which ones should be appended or updated.
      const { appendBatch, updateBatch } = processGetSpreadsheetResponse(response.data, events)

      const promises = [
        processUpdateBatch(mappingSettings, updateBatch, gs),
        processAppendBatch(mappingSettings, appendBatch, gs)
      ]

      return Promise.all(promises)
    })
    .catch((error) => {
      console.log(error)
    })
}

export { processData }
