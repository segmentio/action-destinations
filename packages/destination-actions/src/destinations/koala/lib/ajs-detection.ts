export interface AnalyticsEventContext {
  library?: {
    name?: string
  }
}

export function isAJSEvent(eventContext?: AnalyticsEventContext): boolean {
  return eventContext?.library?.name === 'analytics.js'
}
