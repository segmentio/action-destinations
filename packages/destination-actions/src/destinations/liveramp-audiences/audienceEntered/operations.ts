import { PayloadValidationError, RequestClient } from '@segment/actions-core'
import type { Settings } from '../generated-types'
import type { Payload } from './generated-types'
import { uploadS3, validateS3 } from '../s3'

async function processData(request: RequestClient, settings: Settings, payloads: Payload[]) {
  // STRATCONN-2554: Add support for SFTP
  if (settings.upload_mode == 'S3') {
    validateS3(settings)
  } else {
    throw new PayloadValidationError(`Unrecognized upload mode: ${settings.upload_mode}`)
  }
  // STRATCONN-2584: error if less than 25 elements in payload

  // Prepare header row. Expected format:
  // liveramp_audience_key[1],identifier_data[0..n]
  const rows = []
  const headers = ['audience_key']
  if (payloads[0].identifier_data) {
    for (const identifier of Object.getOwnPropertyNames(payloads[0].identifier_data)) {
      headers.push(identifier)
    }
  }
  rows.push(headers.join(payloads[0].delimiter))

  // Prepare data rows
  for (const payload of payloads) {
    const row = []
    row.push(payload.audience_key)
    if (payload.identifier_data) {
      for (const identifier of Object.getOwnPropertyNames(payload.identifier_data)) {
        row.push(payload.identifier_data[identifier] as string)
      }
    }
    rows.push(row.map(enquoteIdentifier).join(payload.delimiter))
  }

  // STRATCONN-2584: verify multiple emails are handled
  const filename = `${payloads[0].audience_name}_PII_${payloads[0].received_at}.csv`
  const fileContent = rows.join('\n')

  if (settings.upload_mode == 'S3') {
    return await uploadS3(settings, filename, fileContent, request)
  }
}

/*
  To avoid collision with delimeters, we should surround identifiers with quotation marks.
  https://docs.liveramp.com/connect/en/formatting-file-data.html#idm45998667347936

  Examples:
  LCD TV -> "LCD TV"
  LCD TV,50" -> "LCD TV,50"""
*/
function enquoteIdentifier(identifier: string) {
  return `"${identifier.replace(/"/g, '""')}"`
}

export { processData, enquoteIdentifier }
