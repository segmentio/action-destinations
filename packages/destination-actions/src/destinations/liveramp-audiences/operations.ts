import { createHash } from 'crypto'
import type { Payload } from './audienceEntered/generated-types'

/*
Generates the LiveRamp ingestion file. Expected format:
liveramp_audience_key[1],identifier_data[0..n]
*/
function generateFile(payloads: Payload[]) {
  const rows = []
  const headers = ['audience_key']

  // Prepare header row
  if (payloads[0].identifier_data) {
    for (const identifier of Object.getOwnPropertyNames(payloads[0].identifier_data)) {
      headers.push(identifier)
    }
  }

  if (payloads[0].unhashed_identifier_data) {
    for (const identifier of Object.getOwnPropertyNames(payloads[0].unhashed_identifier_data)) {
      headers.push(identifier)
    }
  }
  rows.push(headers.join(payloads[0].delimiter))

  // Prepare data rows
  for (const payload of payloads) {
    const row = []
    row.push(payload.audience_key)
    if (payload.identifier_data) {
      for (const identifier of Object.getOwnPropertyNames(payload.identifier_data)) {
        row.push(payload.identifier_data[identifier] as string)
      }
    }

    if (payload.unhashed_identifier_data) {
      for (const identifier of Object.getOwnPropertyNames(payload.unhashed_identifier_data)) {
        row.push(hash(payload.unhashed_identifier_data[identifier] as string))
      }
    }
    rows.push(row.map(enquoteIdentifier).join(payload.delimiter))
  }

  // TODO: verify multiple emails are handled
  const filename = payloads[0].filename
  const fileContent = Buffer.from(rows.join('\n'))

  return { filename, fileContent }
}

/*
  To avoid collision with delimeters, we should surround identifiers with quotation marks.
  https://docs.liveramp.com/connect/en/formatting-file-data.html#idm45998667347936

  Examples:
  LCD TV -> "LCD TV"
  LCD TV,50" -> "LCD TV,50"""
*/
function enquoteIdentifier(identifier: string) {
  return `"${identifier.replace(/"/g, '""')}"`
}

const hash = (value: string): string => {
  const hash = createHash('sha256')
  hash.update(value)
  return hash.digest('hex')
}

export { generateFile, enquoteIdentifier }
