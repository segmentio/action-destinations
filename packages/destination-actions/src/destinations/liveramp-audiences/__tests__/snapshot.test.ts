import { createTestEvent, createTestIntegration } from '@segment/actions-core'
import { generateTestData } from '../../../lib/test-data'
import destination from '../index'
import nock from 'nock'
import { enquoteIdentifier } from '../operations'

const testDestination = createTestIntegration(destination)
const actionSlug = 'audienceEntered'
const destinationSlug = 'LiverampAudiences'
const seedName = `${destinationSlug}#${actionSlug}`

describe(`Testing snapshot for ${destinationSlug}'s ${actionSlug} destination action:`, () => {
  beforeAll(() => {
    const mockDate = new Date(12345)
    jest.spyOn(global, 'Date').mockImplementation(() => mockDate as unknown as string)
  })

  it('required fields', async () => {
    const action = destination.actions[actionSlug]
    const [eventData, settingsData] = generateTestData(seedName, destination, action, true)
    eventData.delimiter = ','
    settingsData.upload_mode = 'S3'
    nock(/.*/).persist().get(/.*/).reply(200)
    nock(/.*/).persist().post(/.*/).reply(200)
    nock(/.*/).persist().put(/.*/).reply(200)

    const events = new Array(25).fill(0).map(() =>
      createTestEvent({
        properties: eventData
      })
    )

    const responses = await testDestination.testBatchAction(actionSlug, {
      events,
      mapping: events[0].properties,
      settings: settingsData,
      auth: undefined
    })

    const request = responses[0].request
    const rawBody = await request.text()

    try {
      const json = JSON.parse(rawBody)
      expect(json).toMatchSnapshot()
      return
    } catch (err) {
      expect(rawBody).toMatchSnapshot()
    }

    expect(request.headers).toMatchSnapshot()
  })

  it('all fields', async () => {
    const action = destination.actions[actionSlug]
    const [eventData, settingsData] = generateTestData(seedName, destination, action, false)
    eventData.delimiter = ','
    settingsData.upload_mode = 'S3'

    nock(/.*/).persist().get(/.*/).reply(200)
    nock(/.*/).persist().post(/.*/).reply(200)
    nock(/.*/).persist().put(/.*/).reply(200)

    const events = new Array(25).fill(0).map(() =>
      createTestEvent({
        properties: eventData
      })
    )

    const responses = await testDestination.testBatchAction(actionSlug, {
      events,
      mapping: events[0].properties,
      settings: settingsData,
      auth: undefined
    })

    const request = responses[0].request
    const rawBody = await request.text()

    try {
      const json = JSON.parse(rawBody)
      expect(json).toMatchSnapshot()
      return
    } catch (err) {
      expect(rawBody).toMatchSnapshot()
    }
  })

  it('missing minimum payload size', async () => {
    const action = destination.actions[actionSlug]
    const [eventData, settingsData] = generateTestData(seedName, destination, action, false)
    eventData.delimiter = ','
    settingsData.upload_mode = 'S3'

    nock(/.*/).persist().get(/.*/).reply(200)
    nock(/.*/).persist().post(/.*/).reply(200)
    nock(/.*/).persist().put(/.*/).reply(200)

    const event = createTestEvent({
      properties: eventData
    })

    try {
      await testDestination.testBatchAction(actionSlug, {
        events: [event],
        mapping: event.properties,
        settings: settingsData,
        auth: undefined
      })
      throw new Error('expected action to throw')
    } catch (err) {
      expect(err).toMatchSnapshot()
    }
  })

  it('enquotated indentifier data', async () => {
    const identifiers = [`LCD TV,50"`, `"early-bird" special`, `5'8"`]
    const enquotedIdentifiers = identifiers.map(enquoteIdentifier)

    expect(enquotedIdentifiers).toMatchSnapshot()
  })
})
