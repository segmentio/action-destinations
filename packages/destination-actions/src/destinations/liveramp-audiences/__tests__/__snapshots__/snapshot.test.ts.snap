// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing snapshot for LiverampAudiences's audienceEntered destination action: all fields 1`] = `
"audience_key,testType,testType
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\""
`;

exports[`Testing snapshot for LiverampAudiences's audienceEntered destination action: enquotated indentifier data 1`] = `
Array [
  "\\"LCD TV,50\\"\\"\\"",
  "\\"\\"\\"early-bird\\"\\" special\\"",
  "\\"5'8\\"\\"\\"",
]
`;

exports[`Testing snapshot for LiverampAudiences's audienceEntered destination action: missing minimum payload size 1`] = `[PayloadValidationError: received payload count below LiveRamp's ingestion limits. expected: >=25 actual: 1]`;

exports[`Testing snapshot for LiverampAudiences's audienceEntered destination action: required fields 1`] = `
"audience_key,testType,testType
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\"
\\"PywYAY\\",\\"PywYAY\\",\\"7c7ff3a8560d336d69d788505dafaee28b45e77aecd04dd34648e3435865bb38\\""
`;

exports[`Testing snapshot for LiverampAudiences's audienceEntered destination action: required fields 2`] = `
Headers {
  Symbol(map): Object {
    "authorization": Array [
      "AWS4-HMAC-SHA256 Credential=PywYAY/19700101/PywYAY/s3/aws4_request, SignedHeaders=content-length;content-type;host;x-amz-content-sha256;x-amz-date, Signature=3b1c7252eb8f178c58f423003f652a7b8ee011d6e2b9b93891ac9d9bc370b958",
    ],
    "content-length": Array [
      "2155",
    ],
    "content-type": Array [
      "application/x-www-form-urlencoded; charset=utf-8",
    ],
    "host": Array [
      "PywYAY.s3.amazonaws.com",
    ],
    "user-agent": Array [
      "Segment (Actions)",
    ],
    "x-amz-content-sha256": Array [
      "96dc25e0e1fe62a71b555772ecc16e57abf5544c149d8522d9486994d605c3c3",
    ],
    "x-amz-date": Array [
      "19700101T000012Z",
    ],
  },
}
`;
